### DISCUSSION TITLE
Migration Plan Kickoff

### DISCUSSION THREAD

**Discussion by mslipper** [ethereum-optimism/optimistic-specs#242] [Design Decisions]
Hey all! Starting a discussion to begin hammering out the details of how we will transition from Optimism v0 to Bedrock. My goal here is to create a high-level list of things we need to consider for the upgrade. I'll then dig deeper into each one and put together a migration plan.

To kick this off, here are a couple of things that come to mind for me:

## Genesis Configuration

How will we seed Bedrock with Optimism v0's state? In the past we've done this by creating a custom genesis file that gets read by `geth init`. Would this process be the same for Bedrock? Note that we will *not* be performing another regenesis, so I'm curious how this requirement changes the upgrade path.

## Contract Migration

How will we be migrating data/funds from the pre-Bedrock contracts? Are our current proxies sufficient to work with the new contracts, or will we need something different?

## Bedrock Activation

Ideally, we'd deploy Bedrock like a traditional hardfork: at some block height, the Bedrock system would automatically activate and v0 would deactivate. While this might be hard for us to do, I'd like us to at least consider it. It'd dramatically reduce the upgrade burden on our users. Ideally, users could update their libraries once well in advance of the Bedrock release.

Mark's Daisy Chain Proxy might be able to help with this.

## Tooling + Docs

I'm assuming that our existing tooling (e.g., the SDK, our build system, `op-replica`, etc.) and documentation will need to be updated. We should have all this ready well before release in order to mitigate any disruption to existing developers.

What other things am I missing? I don't need a ton of detail - I'd like to go breadth-first, then start adding depth to each thing.

**Comment by maurelian** [comment]
Regarding contract migration, I believe that we will need to make the following changes based on the  [deposit contract](https://github.com/ethereum-optimism/optimistic-specs/blob/main/packages/contracts/contracts/L1/DepositFeed.sol) design, and how I expect the withdrawals design will work out: 

1. Moving ETH from the L1 bridge to the deposit feed contract:
	1. We'll also need to figure out how to manage the transition from OVM_ETH balances on L2 to native ETH balances. This will include [removing the associated logic from the L2 bridge](https://github.com/ethereum-optimism/optimism/blob/4c486d70b852d88eea014a5d1c432dafde1cbd4e/packages/contracts/contracts/L2/messaging/L2StandardBridge.sol#L106). 
2. I believe that we need to review the proxies we're using. Ideally we'd even integrate the L1 Cross Domain messenger into the L1 Bridge, which shouldn't be a huge amount of work, but should be done carefully. 

I hope to soon be able to enumerate these things more clearly, and will update this when I can.

**Reply by mslipper** [reply]
To confirm: the migration from OVM_ETH to native ETH is a requirement for Bedrock? Asking because Mark mentioned this below as well, and this doesn't seem like common knowledge yet.

**Reply by norswap** [reply]
Yes, we have made no special provision in the design for handling ETH via an ERC-20.

**Reply by maurelian** [reply]
The current Deposit and Withdrawal specs include the following fundamental changes to the protocol:

1. We will no longer have `OVM_ETH`.
1. ETH can now be deposited to L2 by sending it to the new `DepositFeed` contract.
1. ETH can now be withdrawn from L2 by sending to the new `WithdrawalsFeed` predeploy contract. (I'm not married to this name, I just think it helps clarify the bilateral relationship between it and the `DepositFeed`.

An exciting feature that this will allow is to deposit ETH from L1 directly into a DeFi protocol on L2.

As a result of those fundamental protocol changes:
1. the ETH in the `L1StandardBridge` will need to be moved into the new `DepositFeed` contract. 
1. The `L2StandardBridge` will no longer be the recipient of an ETH withdrawal.
1. Cross domain messages will now have an additional `value` field.

For backwards compatibility, we could do something where the L1 and L2 bridge contracts can still be used for deposits
and withdrawals of ETH, but just have the bridge contracts forward to the Feed contracts. Using the old bridges
would not have as much in the way of gas savings, and would lack the usefulness of direct cross-domain calls with value.

**Comment by trianglesphere** [comment]
In the rollup node we have a `genesis` where the bedrock system starts sequencing from. It consists of the following
```
L1 Genesis - L1 block after which to start looking for deposits + batches
L2 Genesis - block hash/number in the the rollup node starts building off of
```

We might need a small downtime (order minutes) to configure the rollup genesis, but there's probably something that could be done to import the state into the rollup node and create a genesis block (sigh, I guess this is re-genesis, but then rely on daisy chain). The other option is to pre-populate L2 geth and have it run in the background and then configure the rollup genesis on the rollup node and that's when it will start sequencing/doing everything

**Comment by norswap** [comment]
**Genesis Configuration**

If we're going by the [Update Mechanisms](https://github.com/ethereum-optimism/optimistic-specs/wiki/Update-Mechanisms) page for definitions, the defining thing about re-genesis is whether old blocks are retained.

If we retain old blocks in the execution engine, then it wouldn't be a re-genesis. I'm not sure if re-running `geth init` voids that requirement or not.

Note that we would need to get all the old L2 blocks into the execution engine, and I think there might be dragons in that endeavour? Maybe as long as we guarantee we'll never reorg to before Bedrock's inception, everything is fine? @protolambda 

**Contract Migration**

Concur with John that we need a precise inventory of current contracts, planned contracts, and a mapping between both these things.

**Bedrock Activation**

> Ideally, we'd deploy Bedrock like a traditional hardfork: at some block height, the Bedrock system would automatically activate and v0 would deactivate. 

I think the difficulty with this is that current l2geth and the rollup node (+ execution engine) are almost irreconciliable pieces of software.

I guess some frankenstein plan is possible:
- import both in the same go binary, and somehow handoff from one to the other when reaching a certain block height
    - i.e. run l2geth, when reaching block height X, it simply shuts down to the main routine, which then boots up the rollup-node and the execution engine
- run both current l2geth and the rollup-driver/execution engine at the same time
    - l2geth must be configured to shtudown at block X
    - rollup-driver and execution engine must be busy waiting for block X to occur before booting
    - the execution engine must also somehow keep up to date with the L2 chain (which is it not equipped to process prior to block X) & the chain state
         - so it needs to run block sync + state sync from l2geth, but perform no validation on these things
         - l2geth must keep the sync running after shutting down processing (to avoid race conditions where it shuts down before the sync is done)
         
Note that a "downtime" scenario is pretty much equivalent to the second option above. I think the only added complexity is "live syncing" the chain and the state from l2geth to the execution engine. If we remove the requirement, we could run a quick migration script (does anything need to be done at all if we use the same geth db?) and keep downtime minimal.

**Tooling + Docs**

Besides contract changes, I assume the SDK relies on some assumption of how Optimism works, which may become invalidated with Bedrock. Let's strive to keep the "Bedrock Dev/User Diff" document on notion up-to-date (should we migrate this to the wiki?), then it can serve as a checklist for these things.

**Reply by mslipper** [reply]
To facilitate a hardfork-style activation, I think we could run both `l2geth` and the Bedrock node behind @tynes's daisy chain proxy. We could hardcode a block height in the proxy at which `eth_sendRawTransaction` gets routed to Bedrock.

The proxy is designed to route calls that need historical block data to `l2geth`, so unless I'm missing something I don't think we'd have to import historical blocks into the new execution engine. We would, however, have to run `geth init` to seed Bedrock with the previous chain's state.

**Reply by norswap** [reply]
Good points!

**Reply by trianglesphere** [reply]
I think configuring the rollup + new execution client to start at a specific block is easy. I'm not so sure about configuring the current client to stop. The daisy chain proxy is probably enough to stop the old system (plus maybe a Hardfork style check). Bedrock isn't built to rely on `eth_sendRawTransaction`, but it already has better HF logic.

**Reply by tynes** [reply]
> To facilitate a hardfork-style activation, I think we could run both `l2geth` and the Bedrock node behind @tynes's daisy chain proxy. We could hardcode a block height in the proxy at which `eth_sendRawTransaction` gets routed to Bedrock.
> 
> The proxy is designed to route calls that need historical block data to `l2geth`, so unless I'm missing something I don't think we'd have to import historical blocks into the new execution engine. We would, however, have to run `geth init` to seed Bedrock with the previous chain's state.

See https://github.com/ethereum-optimism/optimistic-specs/discussions/242#discussioncomment-2409206 as to why I don't think having the proxy route write requests is a good idea, it would need additional automation (ie dynamic irregular state transition)

**Comment by tynes** [comment]
I believe we will need to take a snapshot of the state at the final height to feed into `geth init` in the bedrock node. We will also likely want to migrate any ETH ERC20 balances to be actual ETH balances (ie in the account balance field instead of in the ERC20 contract). This will help us reduce the diff. Doing so would be difficult to do a "zero downtime" upgrade, but I think its worth it given the way that this release is simplifying things quite a bit

**Reply by mslipper** [reply]
I agree that we should migrate the ETH ERC20 balances to be actual ETH balances. This will also reduce our attack surface area, and simplify audits.

I assume that the state snapshot makes it hard to do a zero downtime upgrade because we'd have to pause the chain prior to the snapshot?

**Reply by norswap** [reply]
What do these snapshots look like / what mechanism is used to create them?

**Reply by mslipper** [reply]
☝️ I'm curious about this as well, cc @tynes

**Reply by trianglesphere** [reply]
Geth has some DB dump commands. Realistically, if we lean on the proxy idea, we are basically doing a regenesis. The other option is to insert old L2 blocks into the new node, but geth either needs to be able to execute them or skip over them and we don't have code for either (plus it's a high maintenance burden).

**Reply by norswap** [reply]
@trianglesphere Is that also true (the high maintenance burden) if we can guarantee that the old block never need to be executed because of a re-org or RPC call?

Other question, isn't it possible to literally point the new execution engine at the same DB as the old l2geth (for the state)?

**Reply by trianglesphere** [reply]
You could do it that way (include the old state + blocks), but you can't do things like tracing or executing those blocks and need to tell geth that it is unable to do that. Like what if a node tries to sync via those old blocks, it won't be able to execute them so all new nodes doing a sync need to skip them.

It's an interesting idea (and I like the idea of making a mega node that can do it), but it's not zero friction (but nothing is with this problem).

**Reply by tynes** [reply]
> Geth has some DB dump commands. Realistically, if we lean on the proxy idea, we are basically doing a regenesis. The other option is to insert old L2 blocks into the new node, but geth either needs to be able to execute them or skip over them and we don't have code for either (plus it's a high maintenance burden).

Yes the problem here is historical execution, for the debug traces to work it would be a headache of us needing to maintain a larger diff. This is why we have opted towards implementing the daisychain approach of upgrading, where users can opt into running whatever portion of the chain that they need for their application (in regards to older nodes, like they can just not run the oldest nodes if they do not need them)

**Comment by tynes** [comment]
> I assume that the state snapshot makes it hard to do a zero downtime upgrade because we'd have to pause the chain prior to the snapshot?

We could technically implement a dao fork like hardfork that iterates through all accounts and updates the state manually but that would result in a single block consuming an absurd amount of gas, it would be implemented in go so it technically wouldn't be metered but that iteration could take awhile since we are on an older version of geth that doesn't have the state snapshot datastructure.

These snapshots in the past have been generated using `debug_dumpBlock`, see [here](https://github.com/ethereum/go-ethereum/blob/e0761432a43f3c991e870b12b1f49a785202a881/eth/api.go#L273) for the upstream implementation and [here](https://geth.ethereum.org/docs/rpc/ns-debug#debug_dumpblock) for docs. Basically we would commit to a certain blocknumber being the rollover height and then pause sequencer ingress once that height is hit. We would need to write a script that consumes the output of `debug_dumpBlock` that converts the balances of OVM ETH to balances in the accounts themselves. Another way to do this would be to instead of depending on the JSON file, just write a go script that uses the leveldb bindings to the `StateDB` itself and then generate the JSON file as the output.

We would need to determine if there are any other changes to L2 system contracts so they can be included as part of this upgrade. One definite one would be the predeploy for L1 blocknumber, as well as the `OVM_GasPriceOracle` being placed behind a proxy (for more simple upgrades in the future) as well as it pulling the L1 portion of the fee from the L1 attributes contract. This would allow existing applications to maintain backwards compatibility.
